internal/format/base_visitor.go

```diff
@@ -0,0 +1,213 @@
+//func (v *baseVisitor) VisitExtensions(element *proto.Extensions)   {}
+
+func (v *baseVisitor) AddFailure(position scanner.Position, format string, args ...interface{}) {
+	v.Failures = append(v.Failures, &text.Failure{
+		Line:    position.Line,
+		Column:  position.Column,
+		Message: fmt.Sprintf(format, args...),
+	})
+}
+
+func (v *baseVisitor) PWithInlineComment(inlineComment *proto.Comment, args ...interface{}) {
+	if inlineComment == nil || len(inlineComment.Lines) == 0 {
+		v.P(args...)
+		return
+	}
+	// https://github.com/emicklei/proto/commit/5a91db7561a4dedab311f36304fcf0512343a9b1
+	//if inlineComment.Cstyle {
+	//args = append(args, inlineComment.Lines[0])
+	//v.P(args...)
+	//for _, line := range inlineComment.Lines[1:] {
+	//v.P(line)
+	//}
+	//return
+	//}
+	args = append(args, ` //`, cleanCommentLine(inlineComment.Lines[0]))
+	v.P(args...)
+	for i, line := range inlineComment.Lines[1:] {
+		line = cleanCommentLine(line)
+		if line == "" && i != len(inlineComment.Lines)-1 {
+			v.P(`//`)
+		} else {
+			v.P(`//`, line)
+		}
+	}
+}
+
+func (v *baseVisitor) PComment(comment *proto.Comment) {
+	if comment == nil || len(comment.Lines) == 0 {
+		return
+	}
+	// https://github.com/emicklei/proto/commit/5a91db7561a4dedab311f36304fcf0512343a9b1
+	//if comment.Cstyle {
+	//for _, line := range comment.Lines {
+	//v.P(line)
+	//}
+	//return
+	//}
+	// this is weird for now
+	// we always want non-c-style after formatting
+	for i, line := range comment.Lines {
+		line = cleanCommentLine(line)
+		if line == "" && !(i == 0 || i == len(comment.Lines)-1) {
+			v.P(`//`)
+		} else {
+			v.P(`//`, line)
+		}
+	}
+}
+
+func (v *baseVisitor) POptions(isFieldOption bool, options ...*proto.Option) {
+	if len(options) == 0 {
+		return
+	}
+	sort.Slice(options, func(i int, j int) bool { return options[i].Name < options[j].Name })
+	prefix := "option "
+	if isFieldOption {
+		prefix = ""
+	}
+	for i, o := range options {
+		suffix := ";"
+		if isFieldOption {
+			if len(options) > 1 && i != len(options)-1 {
+				suffix = ","
+			} else {
+				suffix = ""
+			}
+		}
+		v.PComment(o.Comment)
+		// TODO: this is a good example of the reasoning for https://github.com/uber/prototool/issues/1
+		if len(o.Constant.Array) == 0 && len(o.Constant.OrderedMap) == 0 {
+			// SourceRepresentation() returns an empty string if the literal is empty
+			// if empty, we do not want to print the key or empty value
+			if source := o.Constant.SourceRepresentation(); source != "" {
+				v.PWithInlineComment(o.InlineComment, prefix, o.Name, ` = `, source, suffix)
+			}
+		} else if len(o.Constant.Array) > 0 { // both Array and OrderedMap should not be set simultaneously, need more followup with emicklei/proto
+			v.Failures = append(
+				v.Failures,
+				text.NewFailuref(o.Position, "INVALID_PROTOBUF", "top-level options should never be arrays, this should not compile with protoc"),
+			)
+		} else { // len(o.Constant.OrderedMap) > 0
```

Comment:
Do we need this conditional comment? Was this necessary before?

Reply:
Yes, you need to figure out if one of the two as length > 0. This is what this `else` is doing, so want to explicitly comment it.